Clean up build warnings

TWiStErRob opened the pull request:
... and prevent future ones.

Didn't want to go straight for adding `-Werror`, because it may clobber local hacking; but I think on CI the commands should include `-Pstrict.compilation=true`.

TWiStErRob committed:
Enable javac lint

TWiStErRob commented on reflect/src/main/java/dagger/reflect/ComponentInvocationHandler.java:
```diff
@@ -75,6 +75,7 @@ private ComponentInvocationHandler(BindingGraph graph) {
             "Members injection methods may only return the injected type or void: " + method);
       }
 
+      @SuppressWarnings({"unchecked", "RedundantCast"})
```
see https://youtrack.jetbrains.com/issue/IDEA-206560

JakeWharton replied:
can you leave this as a comment in source?

TWiStErRob replied:
done

also fixed 2 dangling alignment issues like this:
https://github.com/JakeWharton/dagger-reflect/compare/9230378380e3d07a5ecbb3b3dad3d77b1e24b798..bf291c4220b14819da46e81f9e44c9fe9e9d823c#diff-6f0b84cc3cf4faa16145ef4af80a80daL41

JakeWharton commented:
I've merged e62ad1ba6ef3c6dc52350c95dab10cba84a367f8 and 4f75adaf65d222dabe3c010ee7566dca7f3994be. I'm still thinking about the others. I think bf2 might be obsoleted by a change I'm about to make to bindings and I need to think about 0c4 more. I'm all for more of the error-prone warnings but I'm a bit skeptical of the java ones. Maybe we can turn it on until they annoy me.

TWiStErRob commented:
0c4e911 is unlikely to produce anything other than unchecked, rawtypes and deprecation, which all would be visible in IDEA anyway as warnings, so quick to notice and usually very easy to fix. If "unlikely" is not good enough, it's always an option to flip it around from `-Xlint:all -Xlint:-processing -Xlint:-classfile` to `-Xlint:unchecked -Xlint:rawtypes -Xlint:deprecation`.

TWiStErRob committed:
Fix rawtypes warning in reflect-lint

TWiStErRob committed:
Fix lint warnings in integration test

TWiStErRob committed:
Fix Android Java annotation processor warning in integration test
Fixes :integration-tests:android:javaPreCompileCodegenDebugUnitTest
"Annotation processors must be explicitly declared now.
The following dependencies on the compile classpath are found to contain annotation processor.
Please add them to the testAnnotationProcessor configuration.
See https://developer.android.com/r/tools/annotation-processor-error-message.html for more details."

TWiStErRob committed:
Specify Java compiler encoding explicitly to prevent using system default.
Fixes :integration-tests:upstream:compileTestJava
"Cycles.java:35: error: unmappable character for encoding Cp1252"

TWiStErRob committed:
Fix rawtypes warnings in UnlinkedJustInTimeBinding

TWiStErRob committed:
Fix unchecked warnings in UnlinkedSetBinding by suppressing and explaining

TWiStErRob committed:
Fix unchecked warnings in UnlinkedMapOfProviderBinding by using <?> which is what the link() returns
The two objects in Map<Object, Provider<Object>> would be K and V anyway (as seen in LinkedMapOfValueBinding).
Changed covariant return type to match interface declaration and all other implementations of this interface.

TWiStErRob committed:
Extract generic method to isolate type-safe code

TWiStErRob committed:
Swap annotation and visibility to match other code

TWiStErRob committed:
Extract generic method to isolate type-safe code

TWiStErRob commented:
Rebased and fixed all of the current warnings. `gradlew clean check` runs without warnings, except:

This unfixable warning from AGP:
```
> Task :integration-tests:android:processCodegenDebugResources
Changing the value for a property with a final value has been deprecated. This will fail with an error in Gradle 6.0.
```

and this Errorprone warning which is fixed in #156
```
> Task :reflect:compileJava
P:\projects\contrib\github-dagger2-reflect\reflect\src\main\java\dagger\reflect\Scope.java:82: warning: [ObjectToString] Linker is final and does not override Object.toString, so converting it to a string will print its identity (e.g. `Linker@ 4488aabb`) instead of useful information
.
                + linker); // TODO nice error message with scope chain
                  ^
    (see https://errorprone.info/bugpattern/ObjectToString)
1 warning
```

JakeWharton commented on build.gradle:
```diff
@@ -118,6 +120,24 @@ subprojects {
+        '-Xlint:-auxiliaryclass'
+    ]
+
+    if (project.property("strict.compilation").toBoolean()) {
```
Can we do something like `findProperty("string.compilation") == "true"` to avoid having to change `gradle.properties`?

TWiStErRob replied:
Sure, if you prefer that. The reason behind a default value in gradle.properties was to be explicit; cleanly tell: "this key exists" and "this is the default value". This way multiple usages of the same property can be defaulted by using Gradle's features, instead of potentially duplicating code, just used to doing this nowadays.

JakeWharton reviewed (commented):
Mostly looks good. Just some nits.

TWiStErRob commented on integration-tests/upstream/build.gradle:
```diff
@@ -106,3 +106,7 @@ def ensureDaggerSubmodule = tasks.create('ensureDaggerSubmodule') {
 compileTestJava.dependsOn(ensureDaggerSubmodule)
+
+tasks.withType(JavaCompile).configureEach {
+  options.encoding = 'UTF-8'
```
No, it's not, because of
```
  if (project == project(':integration-tests:upstream')) {
    return
  }
```

TWiStErRob committed:
Fix unchecked warnings in UnlinkedMapOfValueBinding by suppression.
When Scope constructs these objects it gives it the right key.

TWiStErRob commented on reflect/src/main/java/dagger/reflect/UnlinkedMapOfLazyBinding.java:
```diff
@@ -13,6 +13,8 @@
 
   @Override
   public LinkedBinding<?> link(Linker linker, Scope scope) {
+    // assume that mapOfProviderKey is Map<K, Provider<V>> and linker returns the correct Binding
```
Beautiful. Thank you for the link!

TWiStErRob commented:
Double-force-pushed to resolve the conflict (deleted file in another PR) and see the code review changes:
https://github.com/JakeWharton/dagger-reflect/compare/a524045a02c38ec895911264f989ef7d7ea9e3fa..1efaa5293e48123358781e6067cebf83574c3e91

JakeWharton commented on reflect/src/main/java/dagger/reflect/UnlinkedMapOfProviderBinding.java:
```diff
@@ -13,11 +13,10 @@
   }
 
   @Override
-  public LinkedBinding<Map<Object, Provider<Object>>> link(Linker linker, Scope scope) {
```
This is unfortunate. Not that I expect anyone to ever see these types, and they're erased anyway, but it at least forces a bit of safety into the method. Doing `LinkedBinding<Map<?, Provider<?>>>` doesn't work?

JakeWharton commented on build.gradle:
```diff
@@ -118,6 +120,24 @@ subprojects {
+        '-Xlint:-processing',
+        // hide warning: MethodParameters attribute introduced in version 52.0 class files is ignored in version 51.0 class files
+        '-Xlint:-classfile',
+        // hide warning: auxiliary class C in full/path/to/C.java should not be accessed from outside its own source file
```
What triggers all of these? I'm less interested in the description since that can be Googled but more interested in what causes them and why it's safe to ignore. Similar to an `@Ignore` for a test. The 'processing' one has a good justification but not these last two.

TWiStErRob replied:
All 3 are the warning messages, not explanation. I'll research and explain them in the context of this project.

JakeWharton commented:
I've merged and cherry-picked about 7 or 8 of the commits in this PR.

TWiStErRob commented:
> I've merged and cherry-picked about 7 or 8 of the commits in this PR.

ok, thanks; will rebase and continue to discuss the rest

TWiStErRob commented:
Only two relevant commits remain:
 * https://github.com/JakeWharton/dagger-reflect/pull/35/commits/472480b037e07d3c31f552d96d9ff5dee2c25ef0 which would be probably solved by a Gradle/AGP upgrade
 * https://github.com/JakeWharton/dagger-reflect/pull/35/commits/1efaa5293e48123358781e6067cebf83574c3e91 which needs reword based on https://github.com/JakeWharton/dagger-reflect/pull/35#discussion_r310857269

Closing this until the project becomes active again.
